File: Ejercicios_Ordenamiento/Ordenamiento.py
# comparo dos vecinos si estan al reves los intercambio y repito el proceso 
# hast que nadie mas se mueva bubble sort

def bubble_sort(arr):
    n = len(arr)
    comparaciones = 0
    intercambios = 0
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            comparaciones += 1
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                intercambios += 1
                swapped = True
        if not swapped:
            break
    return arr, comparaciones, intercambios
#buscar el más pequeño y ponerlo al 
# principio selection sort

def selection_sort(arr):
    n = len(arr)
    comparaciones= 0
    intercambios = 0
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            comparaciones += 1
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
        intercambios += 1
    return arr, comparaciones, intercambios
# tomo el segubndo y lo inserto donde va 
# respecto al primero insertion sort

def insertion_sort(arr):
    comparaciones = 0
    intercambios = 0
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            comparaciones += 1
            arr[j + 1] = arr[j]
            intercambios += 1
            j -= 1
        arr[j + 1] = key
    return arr, comparaciones, intercambios

# datos: [64, 25, 12, 22, 11, 90, 45, 33]
datos = [64, 25, 12, 22, 11, 90, 45, 33]
for nombre, funcion in [("Bubble", bubble_sort), ("Selection", selection_sort), ("Insertion", insertion_sort)]:
    resultado, comp, interc = funcion(datos.copy())
    print(f"{nombre} sort > {resultado}")
    print(f" Comparaciones: {comp},  Intercambios: {interc}")

File: Ejercicios_Ordenamiento/Ordenamiento.cpp
#include <iostream>
#include <vector>
using namespace std;


void bubbleSort(int arr[], int n, int &comparaciones, int &intercambios) {
    comparaciones = 0;
    intercambios = 0;
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            comparaciones++;
            if (arr[j] > arr[j + 1]) {
                intercambios++;
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

void selectionSort(int arr[], int n, int &comparaciones, int &intercambios) {
    comparaciones = 0;
    intercambios = 0;
    for (int i = 0; i < n - 1; i++) {
        int minIdx = i;
        for (int j = i + 1; j < n; j++) {
            comparaciones++;
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        if (minIdx != i) {
            swap(arr[i], arr[minIdx]);
            intercambios ++;
        }
    }
}


void insertionSort(int arr[], int n, int &comparaciones, int & intercambios) {
    comparaciones = 0;
    intercambios = 0;
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        
       
        while (j >= 0) {
            comparaciones++;
            if (arr[j] > key) {
                arr[j + 1] = arr[j];
                intercambios++;
                j--;
            } else {
                break; 
            }
        }
        arr[j + 1] = key;
    }
}

int main() {
    int arr[] = {64, 34, 25, 12, 22, 11, 90};
    int n = sizeof(arr) / sizeof(arr[0]);
    int comp = 0, inter = 0;

    bubbleSort(arr, n, comp, inter);
    cout << "Bubble Sort - Comparaciones: " << comp << ", Intercambios: " << inter << endl;
    selectionSort(arr, n, comp, inter);
    cout << "Selection Sort - Comparaciones: " << comp << ", Intercambios: " << inter << endl;
    insertionSort(arr, n, comp, inter);
    cout << "Insertio Sort - Comparaciones: " << comp << ", Intercambios: " << inter << endl;
    return 0;
}